### PR TITLE
config/router/tcp_proxy/options: v2 API, boosting and --bootstrap-version CLI removal.

### LINKED ISSUE #17706: Not showing deprecation warnings for boosted protos
It turns out we still attempt to do API boosting in many cases, including bootstrap. This causes protos to load as v2, and once that happens, we miss out on important things like deprecation warnings as the field(s) are not deprecated in v2. This might also apply to PGV validation but I haven't checked that.

After discussing with @htuch we think it would be best to just remove boosting entirely. This might require fixing some tests along the way. If it turns out this is going to take a long time we could potentially:
1) Change the default for do_boosting to false for all functions that support it and audit all prod code to make sure it stays false.
2) Hardcode bootstrap version to 3 and fail if it's set to anything else. (We might still want to do this.)

### PR DESCRIPTION
**Pull request by htuch**
This PR cleans up a bunch of v2 tech debt, including version conversion and API oracle
code:
* Remove all non-trivial field references to hidden_envoy_deprecated_*. These were primarily
  in tcp_proxy, router and Node build_version.
* Remove boosting, hidden type annotation and the API type oracle. We are now only v3 for
  both resource and transport API version.
* Remove --bootstrap-version flag (all bootstraps already must be v3).

Risk level: High (It's possible there were some runtime guarded or not dependencies on
  boosting).
Testing: Validated test functional coverage when removing hidden_envoy_deprecated_* field tests.
Releasing notes: TODO

Fixes #17706

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

**Comment by htuch** [comment]
Still WiP (accidentally didn't create draft), but once I have tests passing, this will be the shape of the solution to #17706 and most of the remaining v2 references in code. CC @mattklein123 @tyxia @alyssawilk @phlax

**Comment by wrowe** [comment]
Need to merge main, @htuch

**Comment by htuch** [comment]
@mattklein123 this should be ready for review now (some small fixups on coverage possible but that's it).

**Review by mattklein123 — APPROVED** [review]
Amazing! So much red! Thank you so much for cleaning all of this up. :tada:

File: source/common/tcp_proxy/tcp_proxy.cc [hunk 1]
```diff
@@ -138,13 +138,6 @@ Config::Config(const envoy::extensions::filters::network::tcp_proxy::v3::TcpProx
     return drain_manager;
   });
 
-  if (config.has_hidden_envoy_deprecated_deprecated_v1()) {
```
Comment (ggreenway) [inline source/common/tcp_proxy/tcp_proxy.cc:141]: I think there may be a bunch of dead code left in tcp_proxy after this, but I can go through and clean it up in a later PR. Very happy to see this going away finally!

**Comment by htuch** [comment]
clang-tidy is timing out. @mattklein123 when you get a chance can you rubber stamp and force submit? Thanks.

File: source/common/tcp_proxy/tcp_proxy.cc (hunk 1 above)
Comment (htuch) [reply source/common/tcp_proxy/tcp_proxy.cc:141]: Sure, please do, this PR is already huge :D

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**